marklogic-client-api/src/main/java/com/marklogic/client/datamovement/impl/QueryBatcherImpl.java

```diff
@@ -127,8 +128,11 @@ public QueryBatcherImpl(Iterator<String> iterator, DataMovementManager moveMgr,
   @Override
   public QueryBatcherImpl onUrisReady(QueryBatchListener listener) {
     if ( listener == null ) throw new IllegalArgumentException("listener must not be null");
-    urisReadyListeners.add(listener);
-    return this;
+    if(resultsSoFar.longValue()< getMaxUris()) {
+        urisReadyListeners.add(listener);
+        return this;
+    }
+    return null;
```

Comment:
I believe the expectation is that listeners will be configured before the job starts.  That is, the resultsSoFar would almost always be zero.  If so, it wouldn't be beneficial to check against the maximum in this method.

Also, fluent method should never return null because chained calls such as withBatchSize() in this example 

    .onUrisReady(...)
    .withBatchSize(...)
 
 would be guaranteed to produce a null pointer exception.